cluster/gce/gci/configure-helper.sh

```diff
@@ -1706,17 +1706,41 @@ function prepare-etcd-manifest {
   if [[ -n "${INITIAL_ETCD_CLUSTER_STATE:-}" ]]; then
     cluster_state="${INITIAL_ETCD_CLUSTER_STATE}"
   fi
+
+  # Configure mTLS for etcd peers.
   if [[ -n "${ETCD_CA_CERT:-}" && -n "${ETCD_PEER_KEY:-}" && -n "${ETCD_PEER_CERT:-}" ]]; then
     etcd_creds=" --peer-trusted-ca-file /etc/srv/kubernetes/etcd-ca.crt --peer-cert-file /etc/srv/kubernetes/etcd-peer.crt --peer-key-file /etc/srv/kubernetes/etcd-peer.key -peer-client-cert-auth "
     etcd_protocol="https"
   fi
 
-  # mTLS should only be enabled for etcd server but not etcd-events. if $1 suffix is empty, it's etcd server.
+  local host_ip=""
+  local etcd_additional_listen_client_urls=""
+  local etcd_additional_listen_metrics_urls=""
+  if [[ ${ETCD_LISTEN_ON_HOST_IP:-} == "true" ]]; then
+    # Override host IP if provided.
+    if [[ -n ${HOST_PRIMARY_IP:-} ]]; then
+      host_ip=${HOST_PRIMARY_IP}
+    else 
+      host_ip=$(hostname -i)
+    fi
+    # Construct additional metrics URL for mTLS-enabled etcd server.
+    if [[ -z "${suffix}" ]]; then
+      etcd_additional_listen_metrics_urls=",http://${host_ip}:2382"
+    fi
+  fi
+
+  # Configure mTLS for clients (e.g. kube-apiserver).
+  # mTLS should only be enabled for etcd server but not etcd-events. If $1 suffix is empty, it's etcd server.
   if [[ -z "${suffix}" && -n "${ETCD_APISERVER_CA_KEY:-}" && -n "${ETCD_APISERVER_CA_CERT:-}" && -n "${ETCD_APISERVER_SERVER_KEY:-}" && -n "${ETCD_APISERVER_SERVER_CERT:-}" && -n "${ETCD_APISERVER_CLIENT_KEY:-}" && -n "${ETCD_APISERVER_CLIENT_CERT:-}" ]]; then
     etcd_apiserver_creds=" --client-cert-auth --trusted-ca-file ${ETCD_APISERVER_CA_CERT_PATH} --cert-file ${ETCD_APISERVER_SERVER_CERT_PATH} --key-file ${ETCD_APISERVER_SERVER_KEY_PATH} "
     etcd_apiserver_protocol="https"
     etcd_livenessprobe_port="2382"
-    etcd_extra_args+=" --listen-metrics-urls=http://${ETCD_LISTEN_CLIENT_IP:-127.0.0.1}:${etcd_livenessprobe_port} "
+    etcd_extra_args+=" --listen-metrics-urls=http://${ETCD_LISTEN_CLIENT_IP:-127.0.0.1}:${etcd_livenessprobe_port}${etcd_additional_listen_metrics_urls:-} "
```

Comment:
Don't you need to the arguments with "," ?

Reply:
Ahh - I see now that the value itself starts with "," - that is pretty misleading...

Reply:
Yeah it's not ideal but to remove the comma from the start will require an additional if control block.

Reply:
Cleaned up the concatenation logic. Now the flag value is default to the original 127.0.0.1 URL, and the host IP URL is appended based on the feature flag.